File: baekjoon/14502.py
import sys, copy
from collections import deque

def dfs(level, now):
    global n, m, result
    
    if level >= 3:
        result = max(result, bfs(copy.deepcopy(now)))
        return
    
    for i in range(n):
        for j in range(m):
            if now[i][j] == 0:
                now[i][j] = 1
                dfs(level + 1, now)
                now[i][j] = 0
                
def bfs(now):
    global dy, dx
    
    for i in range(n):
        for j in range(m):
            if now[i][j] == 2:
                now[i][j] = None
                q = deque()
                q.append((i, j))
                
                while q:
                    now_coor = q.popleft()
                    for k in range(4):
                        next_y = dy[k] + now_coor[0]
                        next_x = dx[k] + now_coor[1]
                        
                        if 0 <= next_y < n and 0 <= next_x < m and now[next_y][next_x] == 0:
                            q.append((next_y, next_x))
                            now[next_y][next_x] = None
    
    return count(now)
    
def count(now):
    count = 0
    
    for i in range(n):
        for j in range(m):
            if now[i][j] == 0:
                count += 1
    return count

if __name__ == '__main__':
    dy = [1, 0, -1, 0]
    dx = [0, 1, 0, -1]
    
    n, m = map(int , sys.stdin.readline().strip().split())
    # 0: 빈칸, 1: 벽, 2: 바이러스
    board = [list(map(int, sys.stdin.readline().strip().split())) for _ in range(n)]
    result = float('-inf')
    
    dfs(0, board)
    print(result)

File: baekjoon/14502.java
import java.util.*;

public class Main {

    private int n;
    private int m;
    private int result;
    private List<List<Integer>> virusList = new ArrayList<>();
    private int[] dy = {1, 0, -1, 0};
    private int[] dx = {0, 1, 0, -1};

    public static void main(String[] args) {
        Main app = new Main();
        app.run();
    }

    private int[][] clone(int[][] array) {
        int[][] newArray = new int[n][m];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                newArray[i][j] = array[i][j];
            }
        }
        return newArray;
    }

    private int bfs(int[][] board) {
        for (List<Integer> virusIndex : virusList) {
            int y = virusIndex.get(0);
            int x = virusIndex.get(1);

            ArrayDeque<List<Integer>> deque = new ArrayDeque<>();
            deque.addLast(Arrays.asList(y, x));
            while (!deque.isEmpty()) {
                List<Integer> now = deque.pollFirst();

                for (int i = 0; i < 4; i++) {
                    int nextY = dy[i] + now.get(0);
                    int nextX = dx[i] + now.get(1);

                    if (nextY >= 0 && nextX >= 0 && nextY < n && nextX < m && board[nextY][nextX] == 0) {
                        board[nextY][nextX] = 2;
                        deque.addLast(Arrays.asList(nextY, nextX));
                    }
                }
            }
        }

        return count(board);
    }

    private int count(int[][] board) {
        int count = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 0) {
                    count++;
                }
            }
        }
        return count;
    }

    private void dfs(int level, int[][] now) {
        if (level >= 3) {
            result = Math.max(result, bfs(clone(now)));
            return;
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (now[i][j] == 0) {
                    now[i][j] = 1;
                    dfs(level + 1, now);
                    now[i][j] = 0;
                }
            }
        }
    }

    private void run() {
        Scanner scan = new Scanner(System.in);
        String[] nm = scan.nextLine().trim().split(" ");
        n = Integer.parseInt(nm[0]);
        m = Integer.parseInt(nm[1]);
        int[][] board = new int[n][m];

        for (int i = 0; i < n; i++) {
            String[] line = scan.nextLine().trim().split(" ");
            for (int j = 0; j < m; j++) {
                board[i][j] = Integer.parseInt(line[j]);
                if (board[i][j] == 2) {
                    virusList.add(Arrays.asList(i, j));
                }
            }
        }

        dfs(0, board);
        System.out.println(result);
    }
}
